### ISSUE TITLE
Grammar.getMaxTokenType() is wrong!

### ISSUE THREAD

**Issue by parrt** [antlr/antlr4#1309]
This is being used for two things: size and max token type. for size it's good. for max token type it is 1 too small!

``` java
    /** How many token types have been allocated so far? */
    public int getMaxTokenType() {
        return typeToTokenList.size() - 1; // don't count 0 (invalid)
    }
```

Note that Recognizer.getTokenTypeMap() should refer to vocab get max type not `getATN().maxTokenType` anyway.

I found this because i couldn't find the display name for last token type via vocabulary.

**Comment by parrt** [comment]
Damn. can't reproduce in test case and can't figure out what was wrong in my grammarian project. I'll close for now.

**Comment by jacekjk** [comment]
I believe there is something wrong with the way tokenTypeMap is constructed in org.antlr.v4.runtime.Recognizer#getTokenTypeMap . The last token in the grammar, that is the token with the maximum token type value is not going to make it into the map.

**Comment by parrt** [comment]
@jacekjk great! Do you have an example?

**Comment by sharwell** [comment]
@parrt @jacekjk I think you're referring to #1042, which was fixed in f13b09eb3d32265939dfa4233112a4655e0cedef.

**Comment by jacekjk** [comment]
@sharwell  you are correct. Thank you.
